### PR TITLE
Ampers Angela P.

### PR DESCRIPTION
**Pull request by AngelaPoland**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I made the decision to allow a user to request a specific room number to book. When assign_available_room method is used, it will assign the requested room first (if available) otherwise it will automatically choose another room without raising an error. I already had Argument errors when no rooms were available or if the room number being requested wasn't 1-20 and I didn't want to kick the user out just because the room they wanted wasn't available. However, using check_availability method first will give all room nums available first as well.
Describe a concept that you gained more clarity on as you worked on this assignment. | Just overall design of a complicated project and having to think through how my initial decisions may affect the future of the project was enlightening and made me really stop to consider possibilities before implementing a certain decision. Also working with and using Date and DateRanges was a trip!
Describe a _nominal test_ that you wrote for this assignment. | That every reservation made is an instance of a reservation class object and that the method also puts that reservation object into an array of all reservations.
Describe an _edge case test_ that you wrote for this assignment. | Raised an error if a date argument given was not an actual Date.new object.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | The more complicated the method, the harder it was to do this in that order. Overall, I really did like this way of psuedocode, test, then write code. Sometimes it was little bits as a time, small psuedo code, one small test, one small amount of code, then repeat a few times even within the same method. Though, the more complicated the method...the more I was drawn to just figuring out the code first before writing the test (beyond the initial test of whatever I wanted the output to be.)

### CODE REVIEW

File: lib/block.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      @num = num
+      @start_date = start_date
+      @end_date = end_date
+      @reserved_rooms = []
+      @cost_per_room = ((end_date - start_date).to_i) * 150.00
+      @full_cost = cost_per_room * num #of full block
+    end
+
+    def range
+      (@start_date...@end_date)
```
Comment (CheezItMan) [inline lib/block.rb:25]: clever

File: lib/block.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    end
+
+    def reserve
+      if reserved_rooms.length < @num
+        reserved_rooms << blocked_rooms[0]
```
Comment (CheezItMan) [inline lib/block.rb:30]: So all the reserved rooms will be the 1st one?

File: lib/reception.rb [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    #returns an array of available rooms based on dates
+    def check_availability(start_date, end_date)
+      available_rooms = []
+
+      @rooms.each do |room|
+        if room.booked_dates.empty? #&& room.blocked_dates.empty?
+          available_rooms << room
+        elsif room.booked_dates.length > 0
+          room.booked_dates.each do |range|
+            if !(range.include? (end_date - 1)) || !(range.include? start_date)
```
Comment (CheezItMan) [inline lib/reception.rb]: What about if `start_date` is before the room's booking range, and `end_date` is after the room's booking range?

File: lib/room.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  NUM_OF_ROOMS = 20
+
+  class Room
+
+    attr_accessor :booked_dates
+    attr_reader :room_num
+
+    def initialize(room_num)
+      @room_num = room_num
+      @booked_dates = [] # should contain elements of Date Ranges
```
Comment (CheezItMan) [inline lib/room.rb]: So a room needs to know about all it's Reservations?  Violation of loose coupling and single responsibility.

File: specs/block_spec.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    it "gives the correct cost" do
+      start_date = Date.new(2018, 3, 7)
+      end_date = Date.new(2018, 3, 10)
+      block = Hotel::Block.new(start_date, end_date, 3)
+      block.full_cost.must_equal 1350.00
+      block.cost_per_room.must_equal 450.0
+    end
+  end
+
+  describe "reserve method" do
```
Comment (CheezItMan) [inline specs/block_spec.rb:37]: You also need to test that it can't include a room in a block if the room has a reservation or at least that no room in a given block has a reservation in that time.

File: specs/reception_spec.rb [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    it "returns an array of rooms" do
+      reception = Hotel::Reception.new
+      reception.all_rooms.must_be_instance_of Array
+      reception.all_rooms[0].must_be_instance_of Hotel::Room
+      reception.all_rooms[19].must_be_instance_of Hotel::Room
+    end
+
+    it "only contains 20 rooms" do
+      reception = Hotel::Reception.new
+      reception.all_rooms.length.wont_be :>, 20
```
Comment (CheezItMan) [inline specs/reception_spec.rb:104]: You could just do, `reception.all_rooms.length.must_equal 20`

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Design** |
Demonstrated classes having a single responsibility |  A good attempt, although having `Room` keep track of reservations is a violation.  
Demonstrated loose coupling |  Moderate coupling, some comments in the code. 
Methods demonstrate a good use of encapsulation, inputs and outputs |   Check, although you should make your helper methods `private`.  
Wave 1 requirements |  Check
Wave 2 requirements |  One bug in your code
Wave 3 requirements |  See my notes
Summary | Not bad, you have some problems in your ability to make reservations and some mixing of roles in your classes.  You also need **much** more indepth testing of your ability to make reservations.  You didn't consider all the cases there.  The same applies to block.  Overall not bad, you've got some work to do, but there's a lot of good work here.  I like your use of ranges for reservations.  Very useful.  Checkout our [reference implementation](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/02-intermediate-ruby/exercises/hotel-revisited.md)
